### PR TITLE
Shrink DefaultConfiguration's memory footprint

### PR DESCRIPTION
**Pull request by jvandort**
Lazily allocate most of its internal state.

Also add internal/experimental flag to trigger heap dump when Gradle detects the daemon is running out of memory

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
Before: 
image

After: 
image

File: platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/health/HealthExpirationStrategy.java [hunk 1]
```diff
@@ -147,16 +165,29 @@ public DaemonExpirationResult checkExpiration() {
                 + "To disable this warning, set '" + DISABLE_PERFORMANCE_LOGGING + "=true'.");
         }
 
+        if (shouldCaptureHeapDump()) {
+            File destinationFile = new File("gradle-heap-dump-" + System.currentTimeMillis() + ".hprof");
```
Comment (big-guy) [inline platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/health/HealthExpirationStrategy.java:169]: We should consider putting this in our daemon specific directory (like we do with logs)

Comment (jvandort) [reply platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/health/HealthExpirationStrategy.java:169]: It actually shows up there already.... Not sure why but it does

**Comment by jvandort** [comment]
Before, the smallest (retained) configuration in the heap dump was 1.9k. Now it is 592B

**Pull request closed without merging**